### PR TITLE
feat: Implement aggregator for agent metrics

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/6724

This PR implements the metrics aggregator to collect internal stats from the agent. Agent metrics are cached in the aggregator until they expire or their values are overwritten.

Changes:
* agent: report stats in every iteration. Stats will include metrics, and these are changing all the time
* agent: report `wgengine` internal metrics, and filter unrelated ones
* coderd: implement metrics aggregator that caches metrics and watches for expired ones
* coderd: use `errGroup` to collect metrics in parallel
* refactor: replace strings with metrics labels (workspace name, agent name, username)
* refactor: rename `log` with `logEntry`

### CODE REVIEW

File: cli/server.go [hunk 1]
```diff
@@ -723,6 +723,17 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
...
+					metricsAggregator, err := prometheusmetrics.NewMetricsAggregator(logger, options.PrometheusRegistry, 0)
+					if err != nil {
+						return xerrors.Errorf("can't initialize metrics aggregator: %w", err)
+					}
+
+					cancelMetricsAggregator := metricsAggregator.Run(ctx)
+					defer cancelMetricsAggregator()
+
+					options.UpdateAgentMetrics = metricsAggregator.Update
+					options.PrometheusRegistry.MustRegister(metricsAggregator)
```
Comment (mafredri) [inline cli/server.go:736]: Is there a non-must method? We could probably give a nicer error than e.g. a panic (assuming that's what it does) if we check it manually.

File: coderd/prometheusmetrics/aggregator.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+	"github.com/coder/coder/codersdk/agentsdk"
+)
+
+const (
+	// MetricHelpForAgent is a help string that replaces all agent metric help
+	// messages. This is because a registry cannot have conflicting
+	// help messages for the same metric in a "gather". If our coder agents are
+	// on different versions, this is a possible scenario.
+	metricHelpForAgent = "Metric is forwarded from workspace agent connected to this instance of coderd."
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:20]:
```suggestion
	metricHelpForAgent = "Metrics are forwarded from workspace agents connected to this instance of coderd."
```

Suggestion: I think it's preferable to use the plural form here.

Side-note: In a HA setup with multiple coderd's, would metrics only be for those few agents connected to that specific coderd instance? Or is it global?

File: coderd/prometheusmetrics/aggregator.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+				for _, m := range ma.queue {
+					desc := prometheus.NewDesc(m.Name, metricHelpForAgent, agentMetricsLabels, nil)
+					valueType, err := asPrometheusValueType(m.Type)
+					if err != nil {
+						ma.log.Error(ctx, "can't convert Prometheus value type", slog.F("name", m.Name), slog.F("type", m.Type), slog.F("value", m.Value), slog.Error(err))
+						continue
+					}
+					constMetric := prometheus.MustNewConstMetric(desc, valueType, m.Value, m.username, m.workspaceName, m.agentName)
+					inputCh <- constMetric
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:155]: I'm not sure how this will be used, but I worry about the use of channels. Essentially the way this behaves now is that it's enough for one misbehaving consumer to lock up this entire goroutine since it's dependent on the collected metrics being consumed in a timely manner.

So essentially we can insert 128 entries here immediately, but if more is queued up then it's on the one who called `Collect(ch)` to ensure things progress smoothly. This could prevent update and cleanup.

File: coderd/workspaceagents.go [hunk 4]
```diff
@@ -1213,39 +1214,51 @@ func (api *API) workspaceAgentReportStats(rw http.ResponseWriter, r *http.Reques
...
+			SessionCountVSCode:          req.SessionCountVSCode,
+			SessionCountJetBrains:       req.SessionCountJetBrains,
+			SessionCountReconnectingPTY: req.SessionCountReconnectingPTY,
+			SessionCountSSH:             req.SessionCountSSH,
+			ConnectionMedianLatencyMS:   req.ConnectionMedianLatencyMS,
+		})
+		return err
+	})
+	errGroup.Go(func() error {
+		return api.Database.UpdateWorkspaceLastUsedAt(ctx, database.UpdateWorkspaceLastUsedAtParams{
```
Comment (mafredri) [inline coderd/workspaceagents.go:1242]: Hmm, why does stats update == workspace used? I imagine it could be started but idle?

File: coderd/workspaceagents.go [hunk 5]
```diff
@@ -1213,39 +1214,51 @@ func (api *API) workspaceAgentReportStats(rw http.ResponseWriter, r *http.Reques
...
+			RxBytes:                     req.RxBytes,
+			TxPackets:                   req.TxPackets,
+			TxBytes:                     req.TxBytes,
+			SessionCountVSCode:          req.SessionCountVSCode,
+			SessionCountJetBrains:       req.SessionCountJetBrains,
+			SessionCountReconnectingPTY: req.SessionCountReconnectingPTY,
+			SessionCountSSH:             req.SessionCountSSH,
+			ConnectionMedianLatencyMS:   req.ConnectionMedianLatencyMS,
+		})
+		return err
```
Comment (mafredri) [inline coderd/workspaceagents.go:1239]: Should we wrap errors to better indicate whence they came from?

**Review by mafredri — APPROVED** [review]
I had some worries with regard to the aggregator main goroutine and potential for lockup, but other than that nothing major. Looks good, left some small suggestions for improvement too. 👍🏻

File: coderd/prometheusmetrics/aggregator.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+				for _, m := range ma.queue {
+					if now.After(m.expiryDate) {
+						hasExpiredMetrics = true
+						break
+					}
+				}
+
+				if hasExpiredMetrics {
+					var j int
+					fresh := make([]annotatedMetric, len(ma.queue))
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:175]:
```suggestion
					fresh := make([]annotatedMetric, 0, len(ma.queue))
```

This caught my eye as I submitted my review, this little change would allow using append and skipping the j++/slice truncation. (And providing capacity would really be optional anyway.)

File: cli/server.go (hunk 1 above)
Comment (mtojek) [reply cli/server.go]: Good call, fixed!

File: coderd/prometheusmetrics/aggregator.go (hunk 2 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: > Suggestion: I think it's preferable to use the plural form here.

Fixed

> Side-note: In a HA setup with multiple coderd's, would metrics only be for those few agents connected to that specific coderd instance? Or is it global?

Only connected ones as metrics are not fetched from a database but just cached in memory. I think that it applies to most of Coder metrics, right?

File: coderd/workspaceagents.go (hunk 4 above)
Comment (mtojek) [reply coderd/workspaceagents.go:1242]: In this PR, I added `errGroup`, but didn't modify the internal logic, so I would rather keep it as is.

File: coderd/prometheusmetrics/aggregator.go (hunk 6 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: Fixed, thanks!

File: coderd/workspaceagents.go (hunk 5 above)
Comment (mtojek) [reply coderd/workspaceagents.go]: Errors are wrapped now.

File: coderd/prometheusmetrics/aggregator.go (hunk 3 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go:155]: > So essentially we can insert 128 entries here immediately, but if more is queued up then it's on the one who called Collect(ch) to ensure things progress smoothly. This could prevent update and cleanup.

128 entries correspond to 128 agents submitting their metrics at the same time. I'm aware of the issue, but considering our scale, I don't think that we should encounter this problem soon. Speaking of a potential solution, do you think that we should introduce a "middle channel" to timeout the `Collect(ch)` if metrics are not streamed? I'm afraid that it may complicate the existing logic.

File: coderd/workspaceagents.go (hunk 4 above)
Comment (mafredri) [reply coderd/workspaceagents.go]: Oh, didn't realize this wasn't added logic. All good then!

File: coderd/prometheusmetrics/aggregator.go (hunk 3 above)
Comment (mafredri) [reply coderd/prometheusmetrics/aggregator.go:155]: Hmm, I see. I think another option, instead of an intermediate channel would be to send a slice over this channel instead of 128 items over the channel. Then Collect can process the slice instead of consuming the channel and the goroutine can immediately continue its work. This way the logic doesn't need to change much at all. This ofc leads to increased memory usage when we exceed 128, but then again, if we do exceed it now it might not be ideal either.

File: coderd/prometheusmetrics/aggregator.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+					}
+				}
+
+				if hasExpiredMetrics {
+					var j int
+					fresh := make([]annotatedMetric, 0, len(ma.queue))
+					for _, m := range ma.queue {
+						if m.expiryDate.After(now) {
+							fresh[j] = m
+							j++
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:179]:
```suggestion
					fresh := make([]annotatedMetric, 0, len(ma.queue))
					for _, m := range ma.queue {
						if m.expiryDate.After(now) {
							fresh = append(fresh, m)
```
I probably explained the make change poorly as this was also a required part of it. A go-ish simplification, if you will. You can do this or then revert the change because assigning `fresh[j]` with `make(..., 0)` would lead to a bad time. 😄

**Review by mafredri — APPROVED** [review]
I think we can continue as-is, or make the channel slice change to avoid lockups in the main goroutine. What I might worry about if we do leave it as-is is how will we notice if it ever becomes a problem, but maybe it won't.

File: coderd/prometheusmetrics/aggregator.go (hunk 7 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: I trust your skills, thanks 👍

File: coderd/prometheusmetrics/aggregator.go (hunk 3 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: > I think another option, instead of an intermediate channel would be to send a slice over this channel instead of 128 items over the channel. Then Collect can process the slice instead of consuming the channel and the goroutine can immediately continue its work. 

In this case `Collect` will need to know when the slice is ready to be "ranged over", otherwise it would be a data race.

I slightly modified your idea. I kept the channel, but I'm passing the entire slice at once, so 128 items limit does not exist anymore. Please let me know your thoughts, @mafredri.

Comment (mafredri) [reply coderd/prometheusmetrics/aggregator.go]: Nice, and it actually looks like how I intended it 😄

File: coderd/prometheusmetrics/aggregator.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+	err = registerer.Register(cleanupHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricsAggregator{
+		log:                    logger,
+		metricsCleanupInterval: metricsCleanupInterval,
+
+		collectCh: make(chan (chan []prometheus.Metric)),
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:99]: I think this will still need a buffer capacity, otherwise we might see the queue full error quite often.

Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: Good catch. I removed it accidentally.
